include/rmm/mr/device/thread_safe_resource_adaptor.hpp

```diff
@@ -0,0 +1,136 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/detail/error.hpp>
+
+#include <mutex>
+
+namespace rmm {
+namespace mr {
+/**
+ * @brief Resource that adapts `Upstream` memory resource adaptor to be thread safe.
+ *
+ * An instance of this resource can be constructured with an existing, upstream resource in order
+ * to satisfy allocation requests. This adaptor wraps allocations and deallocations from Upstream
+ * in a mutex lock.
+ * 
+ * @tparam Upstream Type of the upstream resource used for allocation/deallocation.
+ */
+template <typename Upstream>
+class thread_safe_resource_adaptor final : public device_memory_resource {
+ public:
+
+  using lock_t = std::lock_guard<std::mutex>;
+
+  /**
+   * @brief Construct a new thread safe resource adaptor using `upstream` to satisfy
+   * allocation requests.
+   *
+   * All allocations and frees are protected by a mutex lock
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   *
+   * @param upstream The resource used for allocating/deallocating device memory.
+   */
+  thread_safe_resource_adaptor(Upstream* upstream)
+   : upstream_{upstream} {
+    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
+  }
+
+  /**
+  * @brief Get the upstream memory resource.
+  * 
+  * @return Upstream* pointer to a memory resource object.
+  */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+  * @copydoc rmm::mr::device_memory_resource::supports_streams()
+  */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource with thread safety.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+    lock_t lock(mtx);
+    return upstream_->allocate(bytes, stream);
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to to by `p` and log the
+   * deallocation.
+   *
+   * Every invocation of `logging_resource_adaptor::do_deallocate` will write
```

Comment:
Leftover docs :)

Reply:
Do you otherwise approve? Need to get these RMM PRs moving...

Reply:
Fixed.